feat: Adds shape visitors for local service

ShubhamChaturvedi7 opened the pull request:
*Issue #, if available:*

*Description of changes:*


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Commit:
feat: Adds shape visitors for local service

lucasmcdonald3 commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java:
```diff
@@ -0,0 +1,686 @@
+import software.amazon.smithy.model.traits.EnumTrait;
+import software.amazon.smithy.utils.StringUtils;
+
+// TODO: Remove anonymous function in each of the shape visitor and test if it will work
```
Do we need to do this now or track this? I don't think so, this seems less important, but just checking

ShubhamChaturvedi7 replied:
Nope - this is just for our tracking and GoodToHave.

lucasmcdonald3 commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java:
```diff
@@ -0,0 +1,686 @@
+import software.amazon.smithy.utils.StringUtils;
+
+// TODO: Remove anonymous function in each of the shape visitor and test if it will work
+// TODO: if check with %s to figure out if it's a pointer or not and remove duplicate code when shape is optional vs non optional
```
Should this be a follow-up item? I don't fully understand this one

ShubhamChaturvedi7 replied:
For 36 - There are cases where we create something like:
```
return func() f1 {
return func() f2 {
// Implementation
}()
}()
```

This was because before handling the recursive shapes, we had anonymous function because all type conversion logic was inline, but after adding the support for recursive shapes - we continued using the anonymous function even though now the actual shapes are within their own functions.

For 37: There are bunch of places where we could reduce some of the repeated code-blocks (in Java / codegen) but pushed to future because it requires refactoring and might break the Go code.

lucasmcdonald3 commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java:
```diff
@@ -0,0 +1,686 @@
+    return memberShapeConversionFuncMap.get(shape);
+  }
+
+  protected String referenceStructureShape(final StructureShape shape) {
```
(nit) This is ~100 lines long -- maybe a

```
if serviceShape:
   serviceShape(resourceOrService.asServiceShape)
elif resourceShape:
   resourceShape(resourceOrService.asServiceShape)
else:
    throw ValueError("Unsupported reference shape: " + shape)
```
?

lucasmcdonald3 commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java:
```diff
@@ -0,0 +1,686 @@
+    final Boolean assertionRequired = targetShape.isStructureShape();
+    builder.append(
+      """
+                           var fieldValue %s
```
Is this formatting off or this this the expected indentation?

ShubhamChaturvedi7 replied:
The Go formatter takes care of this, but I've fixed it here as well.

lucasmcdonald3 commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java:
```diff
@@ -0,0 +1,686 @@
+
+  @Override
+  public String timestampShape(final TimestampShape shape) {
+    // TODO: Figure out timestamp types when working on timestampShape
```
I would prefer to raise an exception at codegen time if this is encountered -- safer to make codegen fail on unhandled shapes

ShubhamChaturvedi7 replied:
AFAIK throwing exception fails the codegen on DDBv2 (new model specifically for Go) shapes, hence the `nil` part.

lucasmcdonald3 replied:
Is `DafnyToSmithyShapeVisitor` used for both LocalService and AWS SDK shapes?

(Python only implements TimestampShape for AWS SDKs. It raises an exception at codegen time for LocalServices.)

ShubhamChaturvedi7 replied:
Ahh I should be able to do that. I'll fix it up in a follow up PR since it's going to be again about ShapeVisitors.

lucasmcdonald3 commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java:
```diff
@@ -0,0 +1,686 @@
+      final var targetShape = context
+        .model()
+        .expectShape(memberShape.getTarget());
+      //TODO: Is it ever possible for structure to be nil?
```
One approach to this lets you say the answer is "no":
If the structure is required, it will never be nil
If it is optional, you can check if it is nil before going into the structure converter:

```
nativeOutput = (input.isNotNil ? structureConverter(input) : nil)
```

and only go into the structure converter if it is not nil

But you might need to make a code change to get the optional checking in

ShubhamChaturvedi7 replied:
It is more nuanced than that. A member shape can be Optional but still not be a pointer. Nil check can only operate on pointer types and not on value types. A structure can be value type or pointer type in Go SDK depending on some set of rules - and that set of rules is not very well defined for Dafny vs SDK. Hence the TODO. We added this so that if something fails tomorrow for Structure being nil - we know where to look.

lucasmcdonald3 commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java:
```diff
@@ -0,0 +1,686 @@
+        memberShape.isOptional() &&
+        targetShape.isStructureShape() &&
+        !targetShape.hasTrait(ReferenceTrait.class);
+      final var derivedDataSource =
```
In general I'm struggling to understand what the generated code looks like in context, but this is probably the toughest one for me to grasp.

I think the easiest thing to do is link to some generated code? Do you have an MPL branch I can look at?

I don't know if this works with your timeline, but it would also help if you could leave comments above complicated code blocks with an inline example of what the generated code looks like, and maybe leave some more guiding comments?

ShubhamChaturvedi7 replied:
Yeah it's hard to get a feel for generated code when there's so many template params to be filled in. But adding generated code might pollute the codegen and also end up becoming stale.

Here's a link the MPL that might give you some context: https://github.com/aws/aws-cryptographic-material-providers-library/blob/Golang/dev/AwsCryptographyPrimitives/runtimes/go/ImplementationFromDafny-go/awscryptographyprimitivessmithygenerated/to_native.go#L67

lucasmcdonald3 replied:
Thanks! This is helpful

lucasmcdonald3 commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/SmithyToDafnyShapeVisitor.java:
```diff
@@ -0,0 +1,724 @@
+  }
+
+  @Override
+  public String timestampShape(final TimestampShape shape) {
```
Ditto as in the reverse conversion -- raise exception at codegen time here?

ShubhamChaturvedi7 replied:
Same issue here with the exception.

lucasmcdonald3 replied:
Do we need Timestamps in LocalService?

ShubhamChaturvedi7 replied:
Responded above, will follow up in the next PR.

lucasmcdonald3 commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/ShapeVisitorHelper.java:
```diff
@@ -0,0 +1,151 @@
+    return optionalShapesToDafny.get(shape);
+  }
+
+  public static String toNativeShapeVisitorWriter(
```
I think the methods in this file also need some Javadocs to help out readers.
I know what's going on here since I wrote something similar in Python, but it's probably not clear to anyone outside of you and me (and maybe the Rust folks?).
Plus, this isn't a standard thing in Smithy, so folks coming from Smithy won't know what this is.

Where is this generated? (Just the filename(s) should be fine)
What is it generating and why? (Conversion methods for shapes with members, this solves recursive shapes)

ShubhamChaturvedi7 replied:
Added a class level javadoc explaining this is a delegator and helps avoid infinite recursion while generating typeconversion methods.

Also added relevant method javadocs.

lucasmcdonald3 commented on codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/SmithyToDafnyShapeVisitor.java:
```diff
@@ -0,0 +1,724 @@
+      }
+    }
+
+    throw new UnsupportedOperationException(
```
Yeah, like this. Do this on the reverse function

lucasmcdonald3 reviewed (commented):
The more I digest this the more I like it.

If you could do a once-over to add some single-line comments inside some of the longer functions to help explain what's going on, that would help others understand it more quickly.

I would really like to see some generated code in context before I approve this but it seems reasonable.

Commit:
fix: PR Comments

Commit:
fix: PR Comments

ShubhamChaturvedi7 commented:
> The more I digest this the more I like it.
> 
> If you could do a once-over to add some single-line comments inside some of the longer functions to help explain what's going on, that would help others understand it more quickly.
> 
> I would really like to see some generated code in context before I approve this but it seems reasonable.

Added some comments, renamed some variables to make it self explanatory, and did some better formatting to ease readability - our java formatter is really aggressive and breaks line more than what is generally done in Java. This hinders readability and makes method longer than what it actually should be.

Avoided adding lots of comments since it again gets stale and pollutes code (I favor self explanatory code than comments). Let me know if you think we can do few things better.

lucasmcdonald3 reviewed (approved):
TimestampShape exceptions will be in a new PR
